### ISSUE TITLE
Fix KernelApp

### ISSUE THREAD

**Issue by kevin-bates** [takluyver/jupyter_kernel_mgmt#40]
While testing out the change for PR #39 I discovered that KernelApp (and test_kernelapp) have fallen through the cracks.  This is because the test was [launching the KernelApp from jupyter_client](https://github.com/takluyver/jupyter_kernel_mgmt/blob/master/jupyter_kernel_mgmt/tests/test_kernelapp.py#L15) rather than from jupyter_kernel_mgmt!

When looking into KernelApp, it has quite a few issues.  I've started making the necessary changes, and, for the most part, gotten the test to pass despite the fact that the KernelApp doesn't work completely.

I have pushed those changes [here](https://github.com/gateway-experiments/jupyter_kernel_mgmt/tree/fix-kernelapp), but I'm hoping someone can fill in the blanks.  I don't fully understand the purpose of the application.  For example, it strikes me as odd that `BlockingKernelClient` is created only once KernelApp's shutdown method is called (via signal handlers).

Hmm, I'll go ahead and create a WIP PR with hopes that it can be a collaborative effort.

Thanks.

**Comment by echarles** [comment]
Upon the fix in the source code, we need to ensure jupyter_client is not more installed in the dev/test env for jupyter_kernel_mgmt. Not sure where it was been pulled from?

**Comment by kevin-bates** [comment]
I agree, there must be something that implicitly pulls it in.  Didn't the docs require jupyter_client in some manner?

Here's another case of a jupyter_client reference that would allow the test to pass!  https://github.com/takluyver/jupyter_kernel_mgmt/blob/master/jupyter_kernel_mgmt/tests/test_kernelspec.py#L157-L164

What should we do about KernelspecApp?  It wasn't carried over in the original JKM implementation.  I have [developed a similar app for the remote kernel providers](https://github.com/gateway-experiments/yarn_kernel_provider/blob/master/yarn_kernel_provider/kernelspecapp.py) I've prototyped and it seems like the various providers should provide applications or instructions for how kernel types are created relative to the given provider.

@takluyver - what are your thoughts on this?  I know you wanted to avoid traitlets - which would be required to support such an application (well, you could probably do this w/o traitlets) - but perhaps that avoidance was more relative to the library itself and not surrounding apps like KernelApp and KernelspecApp.

**Comment by kevin-bates** [comment]
Same goes for [`JuptyerConsoleApp`](https://github.com/jupyter/jupyter_client/blob/master/jupyter_client/consoleapp.py) - which is where the `--existing` option comes into play.

There are four applications that reside in `juptyer_client`:  KernelApp, KernelSpecApp, JupyterConsoleApp, and RunApp.  Of these, I view KernelSpecApp as the only one that should definitely reside in this repo - although arguments can be made for the others well.

**Comment by echarles** [comment]
I have the same question for the App to reside in jupyter server (see https://github.com/jupyter/jupyter_server/pull/151).

For kernel management, `KernelSpecApp` is needed. I still see a need to have a kind of client to usage from the CLI.

**Comment by takluyver** [comment]
It's probably ipykernel pulling in jupyter_client in the tests. I hope that eventually, ipykernel would use jupyter_protocol, but that's some way off. We could maybe block importing it in the test suite (ipykernel should be run in a separate process, so it could still import it). The Qt console [has some code to block imports](https://github.com/jupyter/qtconsole/blob/master/qtconsole/qt_loaders.py). That might be a bit drastic, though.

I would prefer to expunge traitlets from the applications in this repository as well. I don't believe it really makes sense to have a full config mechanism for things like `jupyter kernel` - they're more like command line APIs than full user-facing applications. The exception is JupyterConsoleApp, if that lives here, which is a base class for traitlets-using Jupyter console applications (i.e. qtconsole & jupyter_console).

KernelSpecApp should probably also become something like KernelTypeApp, to clarify that there are more ways to define kernel types now.

**Comment by kevin-bates** [comment]
>  It's probably ipykernel pulling in jupyter_client in the tests.

Yep - definitely [ipykernel](https://github.com/ipython/ipykernel/blob/master/setup.py#L93) - thanks.  I think knowing where it's coming from is sufficient.  None of the files in this repo should import it, but we also need to co-exist on systems, so it's really a matter of cleaning these remaining vestiges.

> I would prefer to expunge traitlets from the applications in this repository as well.

Given that the library itself pulls in traitlets (discovery needs it to load app config, restarter uses it for config and logging and about 4 other modules use it to obtain the appropriate logger) I think that's a difficult thing.   I'm okay with trying to rid any embedded applications of traitlets,  but I'm not sure what the benefit of doing that is given its current presence.

> KernelSpecApp should probably also become something like KernelTypeApp...

This makes sense.  I chose the other approach, that KernelSpecApp is an application that creates kernel types of type 'spec'.  I also took this approach in the gateway PoC providers, [Yarn](https://github.com/gateway-experiments/yarn_kernel_provider/blob/master/yarn_kernel_provider/kernelspecapp.py#L45) and [Kubernetes](https://github.com/gateway-experiments/kubernetes_kernel_provider/blob/master/kubernetes_kernel_provider/kernelspecapp.py#L57).  I don't mind changing them (I think the spec-based providers should "say the same story"), but I think an argument could be made to retain the name, but change the description perhaps.  

Regarding the `install-native` option, it looks like this has been deprecated in favor of `ipython kernel install` since jupyter_client 4.0, so this seems a perfect time to drop it out of the mix.

If changing the filename from KernelSpecApp to KernelTypeApp helps clarify things, I'm all for that, but I think saying the application manages kernel types of 'spec' is the right story.

@takluyver - how important is it to you that KernelTypeApp (or whatever) does not use traitlets given the current usages within the library itself?
